src/Controller/ZoomController.php

```diff
@@ -0,0 +1,106 @@
+use League\OAuth2\Client\Token\AccessTokenInterface;
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\RouterInterface;
+
+class ZoomController extends AbstractController
+{
+    private $router;
+    private $zoomClientId;
+    private $zoomClientSecret;
+
+    public function __construct(string $zoomClientId, string $zoomClientSecret, RouterInterface $router)
+    {
+        $this->router = $router;
+        $this->zoomClientId = $zoomClientId;
+        $this->zoomClientSecret = $zoomClientSecret;
+    }
+
+    /**
+     * Ask for Zoom authentication and store the AccessToken in Session.
+     */
+    public function authenticate(Request $request, ZoomApplication $zoomApplication): Response
+    {
+        $session = $request->getSession();
+
+        $provider = new Zoom([
+            'clientId' => $this->zoomClientId,
+            'clientSecret' => $this->zoomClientSecret,
+            'redirectUri' => $this->router->generate('zoom_authenticate', [], RouterInterface::ABSOLUTE_URL),
+        ]);
+
+        if ($request->query->has('error')) {
+            return $this->redirectToRoute('homepage');
+        }
+
+        if (!$request->query->has('code')) {
+            $options = [
+                'scope' => [
+                    'imchat:bot', // from https://marketplace.zoom.us/docs/guides/chatbots/sending-messages
+                    'imchat:write:admin',
+                    'user:read:admin', // needed for getResourceOwner
+                    'contact:read:admin',
+                    //'chat_message:write', For "user" api only, we are "account" level app
+                ],
+            ];
+
+            $authUrl = $provider->getAuthorizationUrl($options);
+
+            $session->set(ZoomApplication::SESSION_KEY_STATE, $provider->getState());
+
+            return new RedirectResponse($authUrl);
+        // Check given state against previously stored one to mitigate CSRF attack
+        } elseif (empty($request->query->get('state')) || ($request->query->get('state') !== $session->get(ZoomApplication::SESSION_KEY_STATE))) {
+            $session->remove(ZoomApplication::SESSION_KEY_STATE);
+
+            throw new AuthenticationException(ZoomApplication::APPLICATION_CODE, 'Invalid OAuth state.');
+        }
+
+        try {
+            /** @var AccessTokenInterface|AccessToken $botToken */
+            $botToken = $provider->getAccessToken('client_credentials');
+
+            /** @var AccessTokenInterface|AccessToken $token */
+            $token = $provider->getAccessToken('authorization_code', [
+                'code' => $request->query->get('code'),
+            ]);
+
+            /** @var ZoomResourceOwner $user */
+            $user = $provider->getResourceOwner($token);
+        } catch (\Exception $e) {
```

Comment:
Is there a better class to catch?

Reply:
If we had different error message for each case (network issue, api down, bad credential) yeah but that's not needed here.